rai/node/node.cpp

```diff
@@ -2742,20 +2865,31 @@ rai::uint128_t rai::election::minimum_threshold (MDB_txn * transaction_a, rai::l
 
 void rai::election::confirm_once (MDB_txn * transaction_a)
 {
-	if (!confirmed.test_and_set ())
+	if (!finished.exchange (true))
 	{
 		auto tally_l (node.ledger.tally (transaction_a, votes));
 		assert (tally_l.size () > 0);
 		auto winner (tally_l.begin ());
 		auto block_l (winner->second);
+		auto node_l (node.shared ());
 		auto exceeded_min_threshold = winner->first > minimum_threshold (transaction_a, node.ledger);
 		if (!(*block_l == *last_winner))
 		{
 			if (exceeded_min_threshold)
 			{
-				auto node_l (node.shared ());
 				node.background ([node_l, block_l]() {
-					node_l->block_processor.process_receive_many (rai::block_processor_item (block_l, true));
+					node_l->block_processor.process_receive_many (rai::block_processor_item (block_l, true), [node_l, block_l](rai::account account, rai::amount amount) {
```

Comment:
Can we avoid this callback? An alternative would be to let the fork block get processed by process_receive_many normally, then let it either trigger the election normally through the block observer, or if it's not in `recent` anymore, use the election history:

```		
                if (this->block_arrival.recent (block_a->hash ())) {
			rai::transaction transaction (store.environment, nullptr, true);
			active.start (transaction, block_a);
		} else if ( ... check election history .. ) {
                        // trigger callback with results of the election + the params from the block observer
                }
```

Reply:
I understand why having a callback there can be misleading, but your proposed solution sounds too convoluted. I'll think about better ways of implementing this.